libvirt/uri/ssh.go

```diff
@@ -91,74 +125,184 @@ func (u *ConnectionURI) dialSSH() (net.Conn, error) {
 		log.Printf("[WARN] Failed to parse ssh config file: %v", err)
 	}
 
-	authMethods := u.parseAuthMethods()
-	if len(authMethods) < 1 {
-		return nil, fmt.Errorf("could not configure SSH authentication methods")
+	// configuration loaded, build tunnel
+	sshClient, err := u.dialHost(u.Host, sshcfg, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// tunnel established, connect to the libvirt unix socket to communicate
+	// e.g. /var/run/libvirt/libvirt-sock
+	address := u.Query().Get("socket")
+	if address == "" {
+		address = defaultUnixSock
+	}
+
+	c, err := sshClient.Dial("unix", address)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to libvirt on the remote host: %w", err)
 	}
+
+	return c, nil
+}
+
+func (u *ConnectionURI) dialHost(target string, sshcfg *ssh_config.Config, depth int) (*ssh.Client, error) {
+
+	if depth > 10 {
+		return nil, fmt.Errorf("[ERROR] dialHost failed: max tunnel depth of 10 reached")
+	}
+
+	log.Printf("[INFO] establishing ssh connection to '%s'", target);
+
 	q := u.Query()
 
+	port := u.Port()
+	if port == "" {
+		port = defaultSSHPort
+	} else {
+		log.Printf("[DEBUG] ssh Port is overriden to: '%s'", port);
+	}
+
+	hostName, err := sshcfg.Get(target, "HostName")
+	if err == nil {
+		if hostName == "" {
+			hostName = target;
+		} else {
+			log.Printf("[DEBUG] HostName is overriden to: '%s'", hostName);
+		}
+	}
+
+	// we must check for knownhosts and verification for each host we connect to.
+	// the query string values have higher precedence to local configs
 	knownHostsPath := q.Get("knownhosts")
 	knownHostsVerify := q.Get("known_hosts_verify")
-	doVerify := q.Get("no_verify") == ""
+	skip_verify := q.Has("no_verify")
```

Comment:
Golang does not use snake-case for variables.

Reply:
roger that. I'll rename to noVerify. The reason I switched from doVerify to skip was because the semantics of the exposed interface were inverted from the internal semantics and it was throwing me off in the debugging (i.e. "do verify" isn't exactly the negation of "no verify" - there are circumstances where there is no verification to be done where "do verify" loses its meaning).